sql/core/src/main/scala/org/apache/spark/sql/execution/command/functions.scala

```diff
@@ -236,6 +236,44 @@ case class ShowFunctionsCommand(
   }
 }
 
+
+/**
+ * A command for users to refresh the persistent function.
+ * The syntax of using this command in SQL is:
+ * {{{
+ *    REFRESH FUNCTION functionName
+ * }}}
+ */
+case class RefreshFunctionCommand(
+    databaseName: Option[String],
+    functionName: String)
+  extends RunnableCommand {
+
+  override def run(sparkSession: SparkSession): Seq[Row] = {
+    val catalog = sparkSession.sessionState.catalog
+    if (FunctionRegistry.builtin.functionExists(FunctionIdentifier(functionName))) {
+      throw new AnalysisException(s"Cannot refresh builtin function $functionName")
+    }
+    if (catalog.isTemporaryFunction(FunctionIdentifier(functionName, databaseName))) {
+      throw new AnalysisException(s"Cannot refresh temporary function $functionName")
+    }
+
+    val identifier = FunctionIdentifier(
+      functionName, Some(databaseName.getOrElse(catalog.getCurrentDatabase)))
+    // we only refresh the permanent function.
+    if (catalog.isPersistentFunction(identifier)) {
+      // register overwrite function.
+      val func = catalog.getFunctionMetadata(identifier)
+      catalog.registerFunction(func, true)
+    } else {
+      // clear cached function.
+      catalog.unregisterFunction(identifier)
```

Comment:
I spent more time thinking about it, and feel like your original proposal is better. When people refreshing a function, they should expect the function to exist and want to use it later. It looks more consistent if we always fail REFRESH TABLE when the function doesn't exist in the catalog anymore.

That said, we can make `unregisterFunction` a noop if the function is not registered, and throw `NoSuchFunctionException` at the end.

Sorry for the back and forth!

Reply:
Thanks for your deep thinking!

Reply:
can you change it? I'd expect something like
```
catalog.unregisterFunction(identifier)
throw new NoSuchFunction...
```